First draft of the first tutorial.

HiPhish opened the pull request:
The PR you wanted.

Commit:
First draft of the first tutorial.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+           ↑
+           k       Hint: The h key is at the left and moves left.
+        ←h   l→          The l key is at the right and moves right.
```
Perhaps this diagram could be removed in favor of the one below. IMO it's counter-productive to compare hjkl with "normal" directional keys. Maybe they could be remembered easier by saying that they operate like a line on an x-y plane:

- j is the leftmost vertical key, and it moves things down
- k is the rightmost vertical key, and it moves things up

That might be too elaborate :D

fmoralesc replied:
Yeah, that is too elaborate. 

I think the analogy with the arrow keys is appropriate to the level of user the tutorial is aimed at, which might be used to move the cursor using the arrow keys. 

Besides, it is unclear in what sense the j and k "move things" up and down. At this point in the tutorial, the user only knows how to move the cursor, using motions in commands is yet ahead...

Pyrohh replied:
yeah, it was a bit much :grimacing:

Still, I don't think we need two diagrams (the flat one is nicer IMO, as it represents the actual key positions relative to each other).

HiPhish replied:
I left the old diagram for comparison, I like the flat one better myself. The cross-shaped diagram looks nicer at first, but nicer is not really my goal. I added a note to the reader why hjkl is superior to arrow keys, so that should do it.

Pyrohh replied:
:+1:

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+      to mode the curse without taking youre hand off. Once getting used to hjkl
+      you will be able to move around much faster. Really!
+
+# Lesson 1.2: EXITING VIM
```
Maybe it's just me, but these all caps titles irk me. I don't think much would be lost if they were made title-case.

HiPhish replied:
Fine by me, I just followed the existing style. I'm not a fan of SHOUTING CASE either, I have even remapped my caps-lock key to `ctrl` OS-wide.

Pyrohh replied:
Excellent, this whole thing is too bold and colorful at the moment :D

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+NOTE: If you are ever unsure about something you typed, press <Esc> to place
+      you in Normal mode.  Then retype the command you wanted.
+
+NOTE: The cursor keys should also work, but these keys have not been chosen
```
I see a lot of these NOTEs... I wonder if they should be worked in to paragraphs, or at least just structured sentences.

HiPhish replied:
Most of these notes were already there, but I agree as well. A note should be really just a side-node, not part of the lesson.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+Vim is a very powerful and popular editor that has many commands. Unlike most
+text editors where you use the keyboard to enter text and arrow keys to move you
+cursor, Vim uses the keyboard to enter commands to manipulate the text.
```
Small note (in case this applies to other places): I think `, in Vim you use ...` makes more sense, as `Vim` isn't the one entering commands.

fmoralesc replied:
Maybe:

> Vim is a very powerful editor, but it is quite different to most. Unlike most text editors where you *must* use the mouse or the arrow keys, Vim has a powerful set of keyboard commands to move around. Likewise, in vim you can use the keyboard to perform many editing actions, like copying, pasting, indenting, etc... (you won't miss anything).`

HiPhish replied:
Yes, Vim does not enter commands, that's stupid. About the other thing, I don't think we need to put focus on Vim being *different*, rather Vim being *better*. How about this?

>These commands allow you to express complex thoughts in just a few
> keystroke to move through text and manipulate it much faster than in most
> text editors.

fmoralesc replied:
:+1:

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+This tutor is designed to describe enough of the commands
+that you will be able to easily use Vim as an all-purpose editor.  It is
+IMPORTANT to remember that this tutor is set up to teach by use.  That means
```
I think bold (if currently possible in vim-tutor-mode) would be a better choice than all-caps; doesn't seem as jarring IMO. I know you said you don't want to polish something that's fundamentally crap; I'm just putting it out there for when you do encounter a need for emphasis on specific words.

fmoralesc replied:
`IMPORTANT` and some other words are highlighted in bold in vim-tutor-mode.

Pyrohh replied:
I think a better scheme than highlighting selected words would just be to surround them with asterisks, like markdown. That way we never need to expand our core list of "bold words" because external projects request so.

If this is already done... oops.

fmoralesc replied:
Oops indeed (with some restrictions).

HiPhish replied:
For me currently when I surround a word with asterisk it disappears completely. I don't think this word needs highlighting anyway, this sentence was that way before I came.

fmoralesc replied:
That is a rule I put in place because allowing `*` can lead to corner cases which are difficult to handle. I've seen way too much of those when working on pandoc markdown's syntax, which is why I decided to be more opinionated here. But I'll make it more flexible.

Pyrohh replied:
I think this calls for a new parser: `llvm-vim-tutor-mode`

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+#             Welcome   to   the   VIM   Tutor
+
+Vim is a very powerful and popular editor that has many commands. Unlike most
```
I don't think its popularity needs to be mentioned.

HiPhish replied:
Yes, that was stupid.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+in Vim is to do tasks in as few steps as possible, which can then be automated.
+Get early into the habit of being lazy.
+
+# Lesson 6.1: THE OPEN COMMAND
```
`6.1` -> `1.6`, and maybe it could be `Lesson 1.6: TEXT EDITING: OPENING` for consistency with the other titles.

HiPhish replied:
Numbers are all over the place anyway.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+word, but in this particular case it acts like the «e» motion. This is
+conscious design choice by the Vim developers, not a bug.
+
+# Lesson something.something: REPLACING TEXT
```
I think it's odd that this doesn't have a subsection (like 1.8) like the other sections do.

fmoralesc replied:
I still doesn't.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+makes no difference, but most instruction use the doubled form.
+
+
+# Lesson something.something: FINDING MOTION
```
Maybe just `find` motion? I'm not sure if `finding` is how it's formally referred to.

HiPhish replied:
Right.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+  5. Type   «P»   to put the line above the cursor.
+
+---> d) Can you learn too? {expect:NULL}
```
@fmoralesc is there any reason for `NULL`? why not just `{expect:}`

fmoralesc replied:
I remember thinking we could need to distinguish between expecting an empty line and the NULL case.

HiPhish replied:
I don't really know the difference between `NULL` and empty.

fmoralesc replied:
NULL is for the case where any contents in the line would be valid. Think "playground".

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+---> c) Intelligence is learned, {expect:NULL}
+---> a) Roses are red, {expect:NULL}
+
+The deleted text is stored in a register and always contains the last deletion
```
Maybe it should be mentioned that registers are text buffers, as we can't assume that a user is familiar with the concept of a CPU register.

fmoralesc replied:
> Maybe it should be mentioned that registers are text buffers

But not use the phrase "text buffer", which can cause confusion. Normally, a "buffer" in vim is the (possibly hidden) content of a window, but registers are *not* buffers in this sense.

Pyrohh replied:
:+1:

HiPhish replied:
Buffer is a no-go, as the term means something very specific in Vim. how about:

> The deleted text is stored in a so-called register, memory holding some text
temporarliy, and always contains the last deletion or yank (see next lesson).

fmoralesc replied:
I like it, but it is a bit ambiguous as to whether registers in general contain the last deletion, or only a  specific one.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+# Summary
+
+asdf
```
:+1:

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+  1. We want to change the prefix all these items. Move the cursor to the first
+     '\'.
+
+  2. Press <C-V> to enter visual mode.
```
Not sure if you wrote this, but IMO this should be `<C-v>`, since while the case of `v` doesn't matter, we don't want users thinking they need to press shift as well.

HiPhish replied:
OK, but I was going with the Vim convention of using the upper-case letter in such cases. I don't really care, we can make it lower-case.

Pyrohh commented:
Thanks for putting the time in to this, but I've really gotta get back to work :P

fmoralesc commented:
@HiPhish  Can you move the tutorial file into the `tutorials/` folder?

Thanks a lot for writing this tutorial, looks nice at first glance :+1:

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+cursor, Vim uses the keyboard to enter commands to manipulate the text.
+There is a great amount of commands, too many to explain in a tutor such as
+this, but as you will see over the course of this tutorial series these commands
+are not arbitrarily chosen, they form a language with its own grammar.
```
> You could say vim's commands form an "editing language" of sorts. This tutorial aims to teach you the basics of it.

HiPhish replied:
If the tutorial did not teach you the basics of the language it would be kind of pointless, wouldn't it? I think that should be expected and doesn't need mentioning.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+Before we continue we need to understand the concept of modal editing first:
+most editors have two modes, a mode where you enter text and a mode where you
+select text.
```
Perhaps it is best to say that many editors are "modeless", and so, radically different from vim. Keys in notepad, for example, don't take modes into account, and will always input characters, whether there is a selection or not (only, if a selection is given, the character will replace the selection).

HiPhish replied:
I disagree. The concept of modes is easier to understand when people have something to compare it to and the two modes of regular editors are like very rudimentary modal editing. When I press a letter key in "normal" mode that letter is inserted before the cursor, but if I press the same key in "visual" mode it will overwrite the selection and place us back into "insert" mode.

fmoralesc replied:
I wouldn't count that as true modes, though, but fair enough ;)

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+---> Roses are red, and tomatoes are red as well. {expect:Roses are red, and tomatoes are red as well.}
+
+An action consits of a count, an operator and a motion, you know that already.
+What I did not tell you is that everything you type in isert mode is part of
```
isert -> insert

HiPhish replied:
Corrected.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+action.
+
+The dot is actually a macro and always contains the last action. We will
+later learn how to define our own macros which are more powerful than that. We
```
define "macros" as an aside. remember: begginers!

HiPhish replied:
Right. I elaborated a bit more.

> The dot is actually a macro, a recorded command, and always contains the last action. We will later learn how to define our own macros which are more powerful than that and can contain any amout of actions. We will also learn how to transplant text properly.
> 
> The dot is still very powerful because it assigns itself automatically and you should use it whenever the possibility arises. Like most other operators it can be combined with a motion.

fmoralesc replied:
:+1: 

I like it, although I'm not sure it's right to say that the dot is a macro, which in vim is more commonly used to refer to the executable registers (used via `q` and `@`).

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+The dot is actually a macro and always contains the last action. We will
+later learn how to define our own macros which are more powerful than that. We
+will also learn how to transplant text properly.
```
what does "transplant text" mean here? should be clearer.

Pyrohh replied:
> what does "transplant text" mean here?

Copy and move I think, I assume you knew the answer to your question :)

fmoralesc replied:
Yes, the question was kinda rethorical ;)

HiPhish replied:
Exactly what transplanting means: you take it from somewhere and place it somewhere else. But it is kind of out of place in this context, so maybe I should remove the last sentence about transplantation?

fmoralesc replied:
Yes, I think it will only lead to confusion here. Regardless the term having that meaning, its use is very unusual when talking about text editing I feel.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+
+# Part 2: EX COMMANDS
+
+** Command mode is where we can execute powerful commands **
```
And which is the mode where we can execute feeble commands? /jk

HiPhish replied:
Now that you say it... Should I remove the powerful?

fmoralesc replied:
Yes, but the issue then is how to make it sound good: "command mode is where we can execute commands" doesn't explain anything...

Pyrohh replied:
Small reminder about this comment.
I think it sounds fine (but plain) if you remove the 
## powerful.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1234 @@
+to write and quit all files.
+
+Unlike actions there is not really a grammar to ex commands, but the commands
+have been chosen in a way that does look like a grammer where it makes sense.
```
There are "micro grammars" for commands: ranges, regex arguments + flags, `|`... Perhaps remove this observation? I feel like it can be more confusing later on than anything.

HiPhish replied:
OK.

Pyrohh commented:
@HiPhish Before you do, please don't rebase/amend your commits yet. It can become difficult to keep track of the changes you've made, and the cleanliness of the history doesn't matter currently, as you can just squash everything before this is merged. I'll try to take a look at everything later (I see a lot of discussion has been had).

Commit:
Refactored until Ex-commands.

Commit:
Finish refactoring first tutorial.

We still have to set the line width to the same value throughout the document
and of course the numbers of the lessons.

Pyrohh commented:
Idk what @fmoralesc thinks, but I'll try to keep my critiques to a minimum, since so much improvements have already been made. It should be easier to make improvements once this is merged, as opposed to perpetually adding on to this PR.

fmoralesc commented:
Sure, let's focus on the surface things before merging, and let's review substance points afterwards. 

Note I don't intend to add this to the first vim-tutor-mode merge into neovim, and I think I'll add it as a separate tutorial first in this repo (however, I'll make it the default so we can more easily test it).

fmoralesc commented:
I'll be focusing on the outline first, see if the sections make sense, etc. I think I'll have a complete review tomorrow.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+Type  :q <Enter>    to close the help window.
+
+You can find help on just about any subject, by giving an argument to the
+":help" command.  Try these (don't forget pressing <Enter>):
```
IMO this is just reinventing `:help` (when run with no arguments) poorly, which I think we should instruct users to run. Specifically, I'm talking about this snippet from `:help`:
```
Get specific help:  It is possible to go directly to whatever you want help
		    on, by giving an argument to the |:help| command.
		    It is possible to further specify the context:
							*help-context*
			  WHAT			PREPEND    EXAMPLE	~
		      Normal mode command      (nothing)   :help x
		      Visual mode command	  v_	   :help v_u
		      Insert mode command	  i_	   :help i_<Esc>
		      Command-line command	  :	   :help :quit
		      Command-line editing	  c_	   :help c_<Del>
		      Vim command argument	  -	   :help -r
		      Option			  '	   :help 'textwidth'

```

HiPhish replied:
Yeah, I wasn't really happy with it either, I just copy-pasted what was already there. Good to see I'm not the only one who thinks that, I'll make is short and to the point.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+
+NOTE: The commands in the lessons will modify the text, but those changes won't
+be saved.  Don't worry about messing things up; just remember that pressing
+[«<Esc>»](<Esc>) and then [u](u) will undo the latest change.
```
@fmoralesc are we still using these Unicode chars for escapes?

fmoralesc replied:
No, we aren't. The syntax for all these spans changed, we'll need to fix that before merging.

Pyrohh replied:
Got it. In that case, is there any reason for those characters being mentioned at https://github.com/fmoralesc/vim-tutor-mode/blob/master/syntax/tutor.vim#L23 ?

fmoralesc replied:
Probably ;)

Pyrohh replied:
Okey, done.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+** To move the cursor, press the «h», «j», «k», «l» keys as indicated. **
+
+        ← ↓ ↑ →    Hint: The h key is at the left and moves left.
+        h j k l          The l key is at the right and moves right.
```
:+1:

HiPhish replied:
I am not sure about the arrows though, they are not ASCII characters, is there any danger they will be messed up for people not using UTF-8 as their encoding? Or can we safely assume that anyone smart enough to want to use Vim is also smart enough to use UTF-8 everywhere?

Pyrohh replied:
> Or can we safely assume that anyone smart enough to want to use Vim is also smart enough to use UTF-8 everywhere?

Maybe that's not a safe assumption... lets keep it for now and see if any issues arise.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+
+Before we continue we need to understand the concept of modal editing first:
+most editors have two "modes", a mode where you enter text and a mode where you
+select text.
```
I think some clarification would be helpful.

`... where you select text, most likely with your mouse.`

HiPhish replied:
Does it really matter what device they use? I assume anyone interested in the power of Vim is already aware of the power of using Shift and the arrow keys to select text quickly and is interested in an even more effective way of editing text.

I think a better clarification would be `where you select text visually`.

Pyrohh replied:
> I think a better clarification would be where you select text visually.

Sure.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+will read to you like a regular sentence instead of a seemingly random
+collection of letters and symbols.
+
+You have already learned four motions so far: h, j, k and l.
```
`already` as well as `so far` is redundant

HiPhish replied:
Is this better?
`So far You have learned four motions: h, j, k and l.`

Pyrohh replied:
sure

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+
+  6. Move on to Lesson 2.5.
+
+Strictly speaking 0 on its own is not a count, it is a motion that moves to the
```
Shouldn't `0` be surrounded be `«0»`?

fmoralesc replied:
Nopes. It should be 

     `0`{normal}

HiPhish replied:
I'll worry about the markup later, along with the proper topic numbering. We can always clean up the little things later once the content is done. Getting hung up on the little stuff now will just slow down the content production.

fmoralesc replied:
:+1: Sure, I was just pointing out what the correct syntax was. Again, we could make things easier by moving drafts like this to something like https://pad.riseup.net/

Pyrohh replied:
@fmoralesc

> Again, we could make things easier by moving drafts like this to something like https://pad.riseup.net/

That might be a good idea.

Anyways, I don't want to get hung up on the small stuff too, but all that's needed is a search and replace.

Pyrohh commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+
+# Summary
+
+asdf
```
perhaps this should be dealt with before merging :grin:

I think it's fine to just remove it for now.

Pyrohh commented:
All done, I tried to keep the comments to a bare minimum. Just waiting on @fmoralesc, and your feedback @HiPhish, then this can probably be merged (I think).

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+There is a great amount of commands, too many to explain in a tutor such as
+this, but as you will see over the course of this tutorial series these
+commands are not arbitrarily chosen, they form a sort of language with its own
+grammar.
```
Note: `there is a great amount of commands, too many to explain...` sounds a bit intimidating, perhaps rephrasing this would be a good idea. Maybe simply remove it and blend with the next paragraph:

>  These commands form a sort of language with its own grammar. This tutorial will help you learn it by introducing you to the set of commands we find is enough to use Vim as an all-purpose editor.

HiPhish replied:
Yes, less is more in a tutorial, especially if it is the very first thing the reader is going to read. The "too many to explain" part was carried over from the previous version, I was being more conservative with my changes.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+same key in insert mode will place a 'j' character in your text.
+
+The next important thing we need to keep in mind is how the syntax of Vim
+commands works: unlike most editor where keyboard commands are just shortcuts,
```
s/editor/editors

HiPhish replied:
Got it.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+
+  5. Repeat steps 2 and 3 with different numbers.
+
+---> This is just a line with words you can move around in. {expect:NULL}
```
`{expect:ANYTHING}`

Pyrohh replied:
Perhaps the syntax could be modified to be something like `{expect:}` or just `{expect}`?
I can't help but feel these long names are influenced by LaTex :grin: 

I prefer the one with the colon, as it's more clear that the argument to `expect` is nil.

fmoralesc replied:
What about `{expect:AnythingReallyWithoutException}`? It really rolls in the tongue. Let me see what I can do, if I remember correctly I did it like this because otherwise we couldn't really expect empty lines, or something like that. It shouldn't be an insurmountable problem...

HiPhish replied:
How about `{expect:Anything really:YES withExceptionOf:Nothing}` ;) Objective-C style all the way.

fmoralesc commented on new_01-vim-beginner.tutor:
```diff
@@ -0,0 +1,1155 @@
+
+# Lesson something.something: WORD-WISE MOTION
+
+** Moving word-wise is faster than character-wise **
```
Perhaps, "For longer mevements, moving word-wise"... It really depends on what you need or want to do.

HiPhish replied:
Nah, the readers are smart enough to know a Vim tutorial is not the Bible. They can figure out on their own when it is more appropriate to move word-wise and when to move character-wise.

fmoralesc replied:
I would still suggest a more neutral tone. In context, this is a value judgement, and it's kinda dumb :p (we should not be less smart than our readers) I would be happier just by replacing the "is" with "can be". Then the lesson teaches *when* it can.

fmoralesc commented:
I'll leave this so we can draft the text together: https://pad.riseup.net/p/krnxUhyJHxZm

I'll be making my review there (mainly, taking care of the markup stuff so @HiPhish doesn't need to worry about it).

HiPhish commented:
One thing I am worried about is the verbosity of this tutorial. While I think this one is more informative than Bram's one, it is undeniable that Bram's tutorial is much quicker to work through. What do you guys think, did I go too far or is it still small enough for a tutorial?

Commit:
Small corrections.

fmoralesc commented:
I think verbose is fine as long as the [readability scores](https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests) don't suffer. We should probably test, though, and even redact like if it was a presentation (aim for a certain time to make it through the tutorial.)

Pyrohh commented:
@HiPhish this was taking too long, so I merged it in a separate branch to make it easier to work on. Thanks for all your work.

HiPhish commented:
Yeah, sorry about that. I find time to work on the tutorial, push my commit and then drop off because of some other task. It's a bad habit, but I'd still like to finish it. Should I write to the pad.riseup page?

fmoralesc commented:
@HiPhish Don't worry ;) We are fixing up the syntax stuff in the `revise-tutorial` branch, you should work over that version from now on. I think that is more or less done, so we can copy that to the pad.riseup page and work there. Any comments, @Pyrohh?

Pyrohh commented:
> Any comments, @Pyrohh?

Nah, copying there sounds fine to me.

fmoralesc commented:
OK, let's do that. https://pad.riseup.net/p/new-vim-tutorial
